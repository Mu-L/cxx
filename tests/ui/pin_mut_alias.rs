mod arg {
    use cxx::ExternType;
    use std::marker::PhantomPinned;

    struct Arg(PhantomPinned);

    unsafe impl ExternType for Arg {
        type Id = cxx::type_id!("Arg");
        type Kind = cxx::kind::Opaque;
    }

    #[cxx::bridge]
    mod ffi {
        unsafe extern "C++" {
            type Arg = crate::arg::Arg;
            fn f(arg: &mut Arg);
        }
    }
}

mod receiver {
    use cxx::ExternType;
    use std::marker::PhantomPinned;

    struct Receiver(PhantomPinned);

    unsafe impl ExternType for Receiver {
        type Id = cxx::type_id!("Receiver");
        type Kind = cxx::kind::Opaque;
    }

    #[cxx::bridge]
    mod ffi {
        unsafe extern "C++" {
            type Receiver = crate::receiver::Receiver;
            fn g(&mut self);
        }
    }
}

mod receiver2 {
    use cxx::ExternType;
    use std::marker::PhantomPinned;

    struct Receiver2(PhantomPinned);

    unsafe impl ExternType for Receiver2 {
        type Id = cxx::type_id!("Receiver2");
        type Kind = cxx::kind::Opaque;
    }

    #[cxx::bridge]
    mod ffi {
        unsafe extern "C++" {
            type Receiver2 = crate::receiver2::Receiver2;
            fn h(self: &mut Receiver2);
        }
    }
}

fn main() {}
