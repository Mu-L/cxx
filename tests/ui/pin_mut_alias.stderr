error[E0271]: type mismatch resolving `<Arg as ExternType>::Kind == Trivial`
  --> tests/ui/pin_mut_alias.rs:15:18
   |
15 |             type Arg = crate::arg::Arg;
   |                  ^^^ type mismatch resolving `<Arg as ExternType>::Kind == Trivial`
   |
note: expected this to be `Trivial`
  --> tests/ui/pin_mut_alias.rs:9:21
   |
 9 |         type Kind = cxx::kind::Opaque;
   |                     ^^^^^^^^^^^^^^^^^
note: required by a bound in `verify_extern_kind`
  --> src/extern_type.rs
   |
   | pub fn verify_extern_kind<T: ExternType<Kind = Kind>, Kind: self::Kind>() {}
   |                                         ^^^^^^^^^^^ required by this bound in `verify_extern_kind`

error[E0271]: type mismatch resolving `<Receiver as ExternType>::Kind == Trivial`
  --> tests/ui/pin_mut_alias.rs:35:18
   |
35 |             type Receiver = crate::receiver::Receiver;
   |                  ^^^^^^^^ type mismatch resolving `<Receiver as ExternType>::Kind == Trivial`
   |
note: expected this to be `Trivial`
  --> tests/ui/pin_mut_alias.rs:29:21
   |
29 |         type Kind = cxx::kind::Opaque;
   |                     ^^^^^^^^^^^^^^^^^
note: required by a bound in `verify_extern_kind`
  --> src/extern_type.rs
   |
   | pub fn verify_extern_kind<T: ExternType<Kind = Kind>, Kind: self::Kind>() {}
   |                                         ^^^^^^^^^^^ required by this bound in `verify_extern_kind`

error[E0271]: type mismatch resolving `<Receiver2 as ExternType>::Kind == Trivial`
  --> tests/ui/pin_mut_alias.rs:55:18
   |
55 |             type Receiver2 = crate::receiver2::Receiver2;
   |                  ^^^^^^^^^ type mismatch resolving `<Receiver2 as ExternType>::Kind == Trivial`
   |
note: expected this to be `Trivial`
  --> tests/ui/pin_mut_alias.rs:49:21
   |
49 |         type Kind = cxx::kind::Opaque;
   |                     ^^^^^^^^^^^^^^^^^
note: required by a bound in `verify_extern_kind`
  --> src/extern_type.rs
   |
   | pub fn verify_extern_kind<T: ExternType<Kind = Kind>, Kind: self::Kind>() {}
   |                                         ^^^^^^^^^^^ required by this bound in `verify_extern_kind`
